nifi-nar-bundles/nifi-prometheus-bundle/nifi-prometheus-reporting-task/src/main/java/org/apache/nifi/reporting/prometheus/PrometheusServer.java

```diff
@@ -0,0 +1,99 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.reporting.prometheus;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+
+import javax.servlet.ServletException;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.nifi.controller.status.ProcessGroupStatus;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.reporting.ReportingContext;
+import org.apache.nifi.reporting.prometheus.api.PrometheusMetricsFactory;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+
+import com.yammer.metrics.core.VirtualMachineMetrics;
+
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.exporter.common.TextFormat;
+
+public class PrometheusServer {
+    private static ComponentLog logger;
+    private Server server;
+    private ServletContextHandler handler;
+
+    public static ReportingContext context;
+    public static boolean sendJvmMetrics;
+    public static String applicationId;
+
+    // set context, SEND_JVM_METRICS, applicationId values are set in onTrigger
+
+    static class MetricsServlet extends HttpServlet {
+        private CollectorRegistry nifiRegistry, jvmRegistry;
+        private ProcessGroupStatus rootGroupStatus;
+
+        @Override
+        protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws ServletException, IOException {
+            logger.info("Do get called");
+
+            rootGroupStatus = context.getEventAccess().getControllerStatus();
+            ServletOutputStream response = resp.getOutputStream();
+            OutputStreamWriter osw = new OutputStreamWriter(response);
+
+            nifiRegistry = PrometheusMetricsFactory.createNifiMetrics(rootGroupStatus, applicationId);
+
+            TextFormat.write004(osw, nifiRegistry.metricFamilySamples());
+
+            if (sendJvmMetrics == true) {
+                jvmRegistry = PrometheusMetricsFactory.createJvmMetrics(VirtualMachineMetrics.getInstance());
+                TextFormat.write004(osw, jvmRegistry.metricFamilySamples());
+            }
+
+            osw.flush();
+            osw.close();
+            response.flush();
+            response.close();
+            resp.setHeader("Content-Type", TextFormat.CONTENT_TYPE_004);
+            resp.setStatus(HttpURLConnection.HTTP_OK);
+            resp.flushBuffer();
+        }
+    }
+
+    public PrometheusServer(InetSocketAddress addr, ComponentLog logger) throws Exception {
+        PrometheusServer.logger = logger;
+        this.server = new Server(addr);
+
+        this.handler = new ServletContextHandler(server, "/metrics");

```

Comment:
Not a huge deal as this runs on its own port, but should this `/metrics` path be configurable?

Reply:
That is the default and every exporter uses the same, so I think it's ok:
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config
https://prometheus.io/docs/instrumenting/writing_exporters/#landing-page